### DISCUSSION TITLE
Feature request: cURL command parsing

### DISCUSSION THREAD

**Discussion by taha-yassine** [darrenburns/posting#73] [Ideas]
It could be very helpful to have the ability to populate the different fields from a cURL command. It could be implemented in different ways but two that I can think of is either a floating pane where the command could be pasted, or allowing direct pasting of the command and detecting if it's a valid one.

**Comment by people-can-fly** [comment]
one of the reason postman is very popular is its import curl feature. it would be nice to have, instead manually typing it

**Reply by ns096** [reply]
I agree, for me that's one of the key features I'm using postman for. And I would love to make the switch to posting

**Comment by MikeBellika** [comment]
For me this is the most important feature. I often use the `Copy as cURL` feature in chrome, so importing these would be great.
@darrenburns are you taking contributions?

PS. Very cool project! Seems very polished and user friendly. I was thinking of building something like this myself, so happy to see that you already did :D

**Reply by darrenburns** [reply]
Yup, I'd accept a PR for curl importing.

Some pointers:

My plan for it was to have it intercept a paste event in the URL bar (override on_paste). So, if the text pasted into the URL bar starts with "curl ", then instead of doing the default Input widget behaviour of inserting the text, Posting would use "post_message" to send a message like "CurlImport" (containing the pasted message) up to the MainScreen, where the string can be parsed into a RequestModal object.

The RequestModal can then be passed into "new_request_modal", so immediately after pasting, the "new request" popup would appear and ask the user for the name/description of the request. I'm pretty sure the machinery for this is all already in place.

The overall result would be that to import from curl, you can just paste the command into the URL bar and supply a name in the modal which appears. I'd probably add a more explicit/obvious approach in the future that would be clear to folks who don't like reading docs, but the URL bar import approach is a good first pass.

**Reply by MikeBellika** [reply]
> Yup, I'd accept a PR for curl importing.
> 
> Some pointers:
> 
> My plan for it was to have it intercept a paste event in the URL bar (override on_paste). So, if the text pasted into the URL bar starts with "curl ", then instead of doing the default Input widget behaviour of inserting the text, Posting would use "post_message" to send a message like "CurlImport" (containing the pasted message) up to the MainScreen, where the string can be parsed into a RequestModal object.
> 
> The RequestModal can then be passed into "new_request_modal", so immediately after pasting, the "new request" popup would appear and ask the user for the name/description of the request. I'm pretty sure the machinery for this is all already in place.
> 
> The overall result would be that to import from curl, you can just paste the command into the URL bar and supply a name in the modal which appears. I'd probably add a more explicit/obvious approach in the future that would be clear to folks who don't like reading docs, but the URL bar import approach is a good first pass.

Seems like a reasonable implementation.
Is there a contribution guide anywhere? Can't seem to get the project running locally

**Reply by darrenburns** [reply]
There's not at the moment I'm afraid. You should be able to install `uv` and then do `uv sync` to install the dependencies, then activate the virtualenv as normal (inside the `.venv` dir).

You can run the tests with `make test`.

**Reply by darrenburns** [reply]
Also, I'm thinking the builtin `argparse` might be a good shout for this 😄

**Reply by MikeBellika** [reply]
https://github.com/user-attachments/assets/6f56357e-83eb-43d7-ac5b-87c33fd27387

Work in progress but got the basics down :D Definitely a bit rusty in Python but getting the hang of it again. 

Are you sure you want the modal on paste? For my use case I would prefer not to name requests. My use case is mainly as a way to take a curl request from chrome and play with it a bit. I don't really need to save requests or name them.
Other programs I've used such as Insomnia or httpie doesn't make you name the request at import time but you can always save them later.

**Reply by darrenburns** [reply]
Awesome! You're right, it's better without the modal :)

**Comment by engdan77** [comment]
This would be of great value 👍🏻. also as a suggestion for something similar in the opposite direction, a way to rapidly generate a cURL command based on the current request, as I think this use case of exchanging queries that way, I might imagine, is quite common.

**Reply by zhangddjs** [reply]
Agree, both import from cURL and output to cURL features are quite important

**Reply by taha-yassine** [reply]
Created https://github.com/darrenburns/posting/discussions/166

**Reply by darrenburns** [reply]
Curl import and export are both supported now (export landed in 2.4.0) :)

Thanks for the great suggestions!
